src/Swarrot/Processor/Retry/RetryProcessor.php

```diff
@@ -23,69 +23,73 @@ public function __construct(ProcessorInterface $processor, MessagePublisherInter
     }
 
     /**
-     * {@inheritDoc}
+     * {@inheritdoc}
      */
     public function process(Message $message, array $options)
     {
         try {
             return $this->processor->process($message, $options);
         } catch (\Exception $e) {
-            $properties = $message->getProperties();
-
-            $attempts = 0;
-            if (isset($properties['headers']['swarrot_retry_attempts'])) {
-                $attempts = $properties['headers']['swarrot_retry_attempts'];
-            }
-            $attempts++;
-
-            if ($attempts > $options['retry_attempts']) {
-                $this->logger and $this->logger->warning(
-                    sprintf(
-                        '[Retry] Stop attempting to process message after %d attempts',
-                        $attempts
-                    ),
-                    [
-                        'swarrot_processor' => 'retry',
-                    ]
-                );
-
-                throw $e;
-            }
-
-            $properties['headers'] = array(
-                'swarrot_retry_attempts' => $attempts,
-            );
-
-            // workaround for https://github.com/pdezwart/php-amqp/issues/170. See https://github.com/swarrot/swarrot/issues/103
-            if (isset($properties['delivery_mode']) && 0 === $properties['delivery_mode']) {
-                unset($properties['delivery_mode']);
-            }
-
-            $message = new Message(
-                $message->getBody(),
-                $properties
-            );
-
-            $key = str_replace('%attempt%', $attempts, $options['retry_key_pattern']);
+            $this->onProcessorException($e, $message, $options);
+        }
+    }
 
+    /**
+     * Retry on exception.
+     * Retries are stopped once retry_attempts has been reached.
+     *
+     * @param \Exception $e
+     * @param Message    $message
+     * @param array      $options
+     */
+    protected function onProcessorException(\Exception $e, Message $message, array $options)
```

Comment:
if you want to handle Throwable on PHP 7 as suggested in #129, the ``\Exception`` typehint should be removed. Otherwise, it is impossible to update the method signature later (changing the signature of a protected method is a BC break)